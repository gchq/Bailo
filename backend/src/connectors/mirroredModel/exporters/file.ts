import { Readable } from 'node:stream'

import { FileWithScanResultsInterface } from '../../../models/File.js'
import { ModelDoc } from '../../../models/Model.js'
import { UserInterface } from '../../../models/User.js'
import { downloadFile, getFileById } from '../../../services/file.js'
import { addEntryToTarGzUpload, initialiseTarGzUpload } from '../../../services/mirroredModel/tarball.js'
import config from '../../../utils/config.js'
import { BadReq, Forbidden, InternalError } from '../../../utils/error.js'
import { FileAction } from '../../authorisation/actions.js'
import authorisation from '../../authorisation/index.js'
import { ScanState } from '../../fileScanning/Base.js'
import scanners from '../../fileScanning/index.js'
import { MirrorKind } from '../index.js'
import { BaseExporter, requiresInit } from './base.js'

export class FileExporter extends BaseExporter {
  protected readonly fileId: string
  protected file: FileWithScanResultsInterface | undefined

  constructor(
    user: UserInterface,
    model: string | ModelDoc,
    file: string | FileWithScanResultsInterface,
    logData?: Record<string, unknown>,
  ) {
    super(user, model, logData)

    if (typeof file === 'string') {
      this.fileId = file
    } else {
      this.file = file
      this.fileId = this.file.id
    }
  }

  getFile() {
    return this.file
  }

  async _init() {
    await super._init()

    if (!this.file) {
      this.file = await getFileById(this.user, this.fileId, this.model)
    }
    const fileAuth = await authorisation.file(this.user, this.model!, this.file, FileAction.Download)
    if (!fileAuth.success) {
      throw Forbidden(fileAuth.info, {
        userDn: this.user.dn,
        modelId: this.modelId,
        fileId: this.fileId,
        ...this.logData,
      })
    }

    if (this.file.size > config.modelMirror.export.maxSize) {
      throw BadReq('Requested export is too large.', {
        size: this.file.size,
        maxSize: config.modelMirror.export.maxSize,
      })
    }

    if (scanners.info()) {
      if (!this.file.avScan || this.file.avScan.length === 0) {
        throw BadReq('The file is missing AV scan(s).', { filename: this.file.name, fileId: this.fileId })
      } else if (this.file.avScan.some((scanResult) => scanResult.state !== ScanState.Complete)) {
        throw BadReq('The file has incomplete AV scan(s).', { filename: this.file.name, fileId: this.fileId })
      } else if (this.file.avScan.some((scanResult) => scanResult.isInfected)) {
        throw BadReq('The file has failed AV scan(s).', { filename: this.file.name, fileId: this.fileId })
      }
    }
  }

  protected getInitialiseTarGzUploadParams(): Parameters<typeof initialiseTarGzUpload> {
    if (!this.model) {
      throw InternalError('Method `getInitialiseTarGzUploadParams` called before `this.model` defined.', {
        ...this.logData,
      })
    }
    return [
      `${this.fileId}.tar.gz`,
      {
        exporter: this.user.dn,
        sourceModelId: this.modelId,
        mirroredModelId: this.model.settings.mirror.destinationModelId!,
        filePath: this.fileId,
        importKind: MirrorKind.File,
      },
      this.logData,
    ]
  }

  @requiresInit
  async addData() {
    // Non-null assertion operator used due to `requiresInit` performing assertion
    await addEntryToTarGzUpload(
      this.tarStream!,
      {
        type: 'stream',
        filename: this.fileId,
        stream: (await downloadFile(this.user, this.fileId)).Body as Readable,
        size: this.file!.size,
      },
      this.logData,
    )
  }
}
