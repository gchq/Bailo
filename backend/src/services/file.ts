import { Schema } from 'mongoose'
import { Readable } from 'stream'

import { getObjectStream, putObjectStream } from '../clients/s3.js'
import { FileAction, ModelAction } from '../connectors/authorisation/actions.js'
import authorisation from '../connectors/authorisation/index.js'
import { FileScanResult, ScanState } from '../connectors/fileScanning/Base.js'
import scanners from '../connectors/fileScanning/index.js'
import FileModel, { FileInterface, FileInterfaceDoc } from '../models/File.js'
import { UserInterface } from '../models/User.js'
import config from '../utils/config.js'
import { BadReq, Forbidden, NotFound } from '../utils/error.js'
import { longId } from '../utils/id.js'
import { plural } from '../utils/string.js'
import log from './log.js'
import { getModelById } from './model.js'
import { removeFileFromReleases } from './release.js'

export function isFileInterfaceDoc(data: unknown): data is FileInterfaceDoc {
  if (typeof data !== 'object' || data === null) {
    return false
  }

  if (
    !('modelId' in data) ||
    !('name' in data) ||
    !('size' in data) ||
    !('mime' in data) ||
    !('bucket' in data) ||
    !('path' in data) ||
    !('complete' in data) ||
    !('avScan' in data) ||
    !('deleted' in data) ||
    !('createdAt' in data) ||
    !('updatedAt' in data) ||
    !('_id' in data)
  ) {
    return false
  }
  return true
}

export async function uploadFile(user: UserInterface, modelId: string, name: string, mime: string, stream: Readable) {
  const model = await getModelById(user, modelId)
  if (model.settings.mirror.sourceModelId) {
    throw BadReq(`Cannot upload files to a mirrored model.`)
  }

  const fileId = longId()

  const bucket = config.s3.buckets.uploads
  const path = `beta/model/${modelId}/files/${fileId}`

  const file = new FileModel({ modelId, name, mime, bucket, path, complete: true })

  const auth = await authorisation.file(user, model, file, FileAction.Upload)
  if (!auth.success) {
    throw Forbidden(auth.info, { userDn: user.dn, fileId: file._id })
  }

  const { fileSize } = await putObjectStream(bucket, path, stream)
  if (fileSize === 0) {
    throw BadReq(`Could not upload ${file.name} as it is an empty file.`, { file: file })
  }
  file.size = fileSize

  await file.save()

  if (scanners.info() && fileSize > 0) {
    const resultsInprogress: FileScanResult[] = scanners.info().map((scannerName) => ({
      toolName: scannerName,
      state: ScanState.InProgress,
      lastRunAt: new Date(),
    }))
    await updateFileWithResults(file._id, resultsInprogress)
    scanners.scan(file).then((resultsArray) => updateFileWithResults(file._id, resultsArray))
  }

  return file
}

async function updateFileWithResults(_id: Schema.Types.ObjectId, results: FileScanResult[]) {
  for (const result of results) {
    const updateExistingResult = await FileModel.updateOne(
      { _id, 'avScan.toolName': result.toolName },
      {
        $set: { 'avScan.$': { ...result } },
      },
    )
    if (updateExistingResult.modifiedCount === 0) {
      await FileModel.updateOne(
        { _id, avScan: { $exists: true } },
        {
          $push: { avScan: { toolName: result.toolName, state: result.state, lastRunAt: new Date() } },
        },
      )
    }
  }
}

export async function downloadFile(user: UserInterface, fileId: string, range?: { start: number; end: number }) {
  const file = await getFileById(user, fileId)
  const model = await getModelById(user, file.modelId)

  const auth = await authorisation.file(user, model, file, FileAction.Download)
  if (!auth.success) {
    throw Forbidden(auth.info, { userDn: user.dn, fileId })
  }

  return getObjectStream(file.bucket, file.path, range)
}

export async function getFileById(user: UserInterface, fileId: string) {
  const file = await FileModel.findOne({
    _id: fileId,
  })

  if (!file) {
    throw NotFound(`The requested file was not found.`, { fileId })
  }

  const model = await getModelById(user, file.modelId)
  const auth = await authorisation.file(user, model, file, FileAction.View)
  if (!auth.success) {
    throw Forbidden(auth.info, { userDn: user.dn, fileId })
  }

  return file
}

export async function getFilesByModel(user: UserInterface, modelId: string) {
  const model = await getModelById(user, modelId)
  const files = await FileModel.find({ modelId })

  const auths = await authorisation.files(user, model, files, FileAction.View)
  return files.filter((_, i) => auths[i].success)
}

export async function getFilesByIds(user: UserInterface, modelId: string, fileIds: string[]) {
  const model = await getModelById(user, modelId)
  if (fileIds.length === 0) {
    return []
  }
  const files = await FileModel.find({ _id: { $in: fileIds } })

  if (files.length !== fileIds.length) {
    const notFoundFileIds = fileIds.filter((id) => files.some((file) => file._id.toString() === id))
    throw NotFound(`The requested files were not found.`, { fileIds: notFoundFileIds })
  }

  const auths = await authorisation.files(user, model, files, FileAction.View)
  return files.filter((_, i) => auths[i].success)
}

export async function removeFile(user: UserInterface, modelId: string, fileId: string) {
  const model = await getModelById(user, modelId)
  const file = await getFileById(user, fileId)
  if (model.settings.mirror.sourceModelId) {
    throw BadReq(`Cannot remove file from a mirrored model`)
  }

  const auth = await authorisation.file(user, model, file, FileAction.Delete)
  if (!auth.success) {
    throw Forbidden(auth.info, { userDn: user.dn })
  }

  await removeFileFromReleases(user, model, fileId)

  // We don't actually remove the file from storage, we only hide all
  // references to it.  This makes the file not visible to the user.
  await file.delete()

  return file
}

export async function getTotalFileSize(fileIds: string[]) {
  const totalSize = await FileModel.aggregate()
    .match({ _id: { $in: fileIds } })
    .group({
      _id: null,
      totalSize: { $sum: '$size' },
    })
  return totalSize.at(0).totalSize
}

export async function markFileAsCompleteAfterImport(path: string) {
  const file = await FileModel.findOneAndUpdate(
    {
      path,
    },
    { complete: true },
  )

  if (!file) {
    log.debug({ path }, 'No file document yet exists for this imported file.')
  }
}

function fileScanDelay(file: FileInterface): number {
  const delay = config.connectors.fileScanners.retryDelayInMinutes
  if (delay === undefined) {
    return 0
  }
  let minutesBeforeRetrying = 0
  for (const scanResult of file.avScan) {
    const delayInMilliseconds = delay * 60000
    const scanTimeAtLimit = scanResult.lastRunAt.getTime() + delayInMilliseconds
    if (scanTimeAtLimit > new Date().getTime()) {
      minutesBeforeRetrying = scanTimeAtLimit - new Date().getTime()
      break
    }
  }
  return Math.round(minutesBeforeRetrying / 60000)
}

export async function rerunFileScan(user: UserInterface, modelId, fileId: string) {
  const model = await getModelById(user, modelId)
  if (!model) {
    throw BadReq('Cannot find requested model', { modelId: modelId })
  }
  const file = await getFileById(user, fileId)
  if (!file) {
    throw BadReq('Cannot find requested file', { modelId: modelId, fileId: fileId })
  }
  const rerunFileScanAuth = await authorisation.file(user, modelId, file, FileAction.Update)
  if (!rerunFileScanAuth.success) {
    throw Forbidden(rerunFileScanAuth.info, { userDn: user.dn, modelId, file })
  }
  if (!file.size || file.size === 0) {
    throw BadReq('Cannot run scan on an empty file')
  }
  const minutesBeforeRescanning = fileScanDelay(file)
  if (minutesBeforeRescanning > 0) {
    throw BadReq(`Please wait ${plural(minutesBeforeRescanning, 'minute')} before attempting a rescan ${file.name}`)
  }
  const auth = await authorisation.model(user, model, ModelAction.Update)
  if (!auth.success) {
    throw Forbidden(auth.info, { userDn: user.dn })
  }
  if (scanners.info()) {
    const resultsInprogress = scanners.info().map((scannerName) => ({
      toolName: scannerName,
      state: ScanState.InProgress,
      lastRunAt: new Date(),
    }))
    await updateFileWithResults(file._id, resultsInprogress)
    scanners.scan(file).then((resultsArray) => updateFileWithResults(file._id, resultsArray))
  }
  return `Scan started for ${file.name}`
}

export async function saveImportedFile(file: FileInterface) {
  await FileModel.findOneAndUpdate({ modelId: file.modelId, _id: file._id }, file, {
    upsert: true,
  })
}
