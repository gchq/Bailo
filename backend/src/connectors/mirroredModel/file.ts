import { PassThrough, Readable } from 'node:stream'

import { Headers } from 'tar-stream'

import { putObjectStream } from '../../clients/s3.js'
import { ScanState } from '../../connectors/fileScanning/Base.js'
import scanners from '../../connectors/fileScanning/index.js'
import FileModel, { FileWithScanResultsInterface } from '../../models/File.js'
import { ModelDoc } from '../../models/Model.js'
import { UserInterface } from '../../models/User.js'
import { createFilePath, downloadFile, getFileById, markFileAsCompleteAfterImport } from '../../services/file.js'
import log from '../../services/log.js'
import { addEntryToTarGzUpload, initialiseTarGzUpload } from '../../services/mirroredModel/tarball.js'
import config from '../../utils/config.js'
import { BadReq, Forbidden, InternalError } from '../../utils/error.js'
import { FileAction } from '../authorisation/actions.js'
import authorisation from '../authorisation/index.js'
import { BaseExporter, BaseImporter, BaseMirrorMetadata, requiresInit } from './base.js'
import { MirrorKind, MirrorKindKeys } from './index.js'

export type FileMirrorMetadata = BaseMirrorMetadata & { importKind: MirrorKindKeys<'File'>; filePath: string }
export type FileMirrorInformation = {
  metadata: FileMirrorMetadata
  sourcePath: string
  newPath: string
}

export class FileExporter extends BaseExporter {
  protected readonly fileId: string
  protected file: FileWithScanResultsInterface | undefined

  constructor(
    user: UserInterface,
    model: string | ModelDoc,
    file: string | FileWithScanResultsInterface,
    logData?: Record<string, unknown>,
  ) {
    super(user, model, logData)

    if (typeof file === 'string') {
      this.fileId = file
    } else {
      this.file = file
      this.fileId = this.file.id
    }
  }

  getFile() {
    return this.file
  }

  async _init() {
    await super._init()

    if (!this.file) {
      this.file = await getFileById(this.user, this.fileId, this.model)
    }
    const fileAuth = await authorisation.file(this.user, this.model!, this.file, FileAction.Download)
    if (!fileAuth.success) {
      throw Forbidden(fileAuth.info, {
        userDn: this.user.dn,
        modelId: this.modelId,
        fileId: this.fileId,
        ...this.logData,
      })
    }

    if (this.file.size > config.modelMirror.export.maxSize) {
      throw BadReq('Requested export is too large.', {
        size: this.file.size,
        maxSize: config.modelMirror.export.maxSize,
      })
    }

    if (scanners.info()) {
      if (!this.file.avScan || this.file.avScan.length === 0) {
        throw BadReq('The file is missing AV scan(s).', { filename: this.file.name, fileId: this.fileId })
      } else if (this.file.avScan.some((scanResult) => scanResult.state !== ScanState.Complete)) {
        throw BadReq('The file has incomplete AV scan(s).', { filename: this.file.name, fileId: this.fileId })
      } else if (this.file.avScan.some((scanResult) => scanResult.isInfected)) {
        throw BadReq('The file has failed AV scan(s).', { filename: this.file.name, fileId: this.fileId })
      }
    }
  }

  protected getInitialiseTarGzUploadParams(): Parameters<typeof initialiseTarGzUpload> {
    if (!this.model) {
      throw InternalError('Method `getInitialiseTarGzUploadParams` called before `this.model` defined.', {
        ...this.logData,
      })
    }
    return [
      `${this.fileId}.tar.gz`,
      {
        exporter: this.user.dn,
        sourceModelId: this.modelId,
        mirroredModelId: this.model.settings.mirror.destinationModelId!,
        filePath: this.fileId,
        importKind: MirrorKind.File,
      },
      this.logData,
    ]
  }

  @requiresInit
  async addData() {
    // Non-null assertion operator used due to `requiresInit` performing assertion
    await addEntryToTarGzUpload(
      this.tarStream!,
      {
        type: 'stream',
        filename: this.fileId,
        stream: (await downloadFile(this.user, this.fileId)).Body as Readable,
        size: this.file!.size,
      },
      this.logData,
    )
  }
}

export class FileImporter extends BaseImporter {
  declare protected metadata: FileMirrorMetadata

  protected bucket: string
  protected updatedPath: string
  protected extractedFile: boolean = false

  constructor(metadata: FileMirrorMetadata, logData?: Record<string, unknown>) {
    super(metadata, logData)
    if (this.metadata.importKind !== MirrorKind.File) {
      throw InternalError('Cannot parse compressed File: incorrect metadata specified.', {
        metadata: this.metadata,
        ...this.logData,
      })
    }
    this.bucket = config.s3.buckets.uploads
    this.updatedPath = createFilePath(this.metadata.mirroredModelId, this.metadata.filePath)
  }

  async processEntry(entry: Headers, stream: PassThrough) {
    if (entry.type === 'file') {
      // Process file
      if (this.extractedFile) {
        throw InternalError('Cannot parse compressed file: multiple files found.', {
          mirroredModelId: this.metadata.mirroredModelId,
          fileId: this.metadata.filePath,
          entry,
          ...this.logData,
        })
      } else {
        const foundFile = await FileModel.findOne({ path: this.updatedPath, complete: true })
        if (foundFile) {
          log.debug(
            { bucket: this.bucket, path: this.updatedPath, ...this.logData },
            'Skipping imported file as it has already been uploaded to S3.',
          )
          // auto-drain the stream
          stream.resume()
        } else {
          await putObjectStream(this.updatedPath, stream, this.bucket)
          await markFileAsCompleteAfterImport(this.updatedPath)
          log.debug(
            { bucket: this.bucket, path: this.updatedPath, name: entry.name, ...this.logData },
            'Imported file successfully uploaded to S3.',
          )
        }
        this.extractedFile = true
      }
    } else {
      // skip entry of type: link | symlink | directory | block-device | character-device | fifo | contiguous-file
      log.debug({ name: entry.name, type: entry.type, ...this.logData }, 'Skipping non-file entry.')
    }
  }

  // Type resolve
  finishListener(resolve: (reason?: FileMirrorInformation) => void, _reject: (reason?: unknown) => void) {
    super.finishListener(resolve, _reject)
  }
}
