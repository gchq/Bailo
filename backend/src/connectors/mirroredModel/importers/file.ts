import { PassThrough } from 'node:stream'

import { Headers } from 'tar-stream'

import { putObjectStream } from '../../../clients/s3.js'
import FileModel from '../../../models/File.js'
import { createFilePath, markFileAsCompleteAfterImport } from '../../../services/file.js'
import log from '../../../services/log.js'
import config from '../../../utils/config.js'
import { InternalError } from '../../../utils/error.js'
import { MirrorKind, MirrorKindKeys } from '../index.js'
import { BaseImporter, BaseMirrorMetadata } from './base.js'

export type FileMirrorMetadata = BaseMirrorMetadata & { importKind: MirrorKindKeys<'File'>; filePath: string }
export type FileMirrorInformation = {
  metadata: FileMirrorMetadata
  sourcePath: string
  newPath: string
}

export class FileImporter extends BaseImporter {
  declare protected readonly metadata: FileMirrorMetadata

  protected readonly bucket: string
  protected readonly updatedPath: string

  protected extractedFile: boolean = false

  constructor(metadata: FileMirrorMetadata, logData?: Record<string, unknown>) {
    super(metadata, logData)
    if (this.metadata.importKind !== MirrorKind.File) {
      throw InternalError('Cannot parse compressed File: incorrect metadata specified.', {
        metadata: this.metadata,
        ...this.logData,
      })
    }
    this.bucket = config.s3.buckets.uploads
    this.updatedPath = createFilePath(this.metadata.mirroredModelId, this.metadata.filePath)
  }

  async processEntry(entry: Headers, stream: PassThrough) {
    if (entry.type === 'file') {
      // Process file
      if (this.extractedFile) {
        throw InternalError('Cannot parse compressed file: multiple files found.', {
          mirroredModelId: this.metadata.mirroredModelId,
          fileId: this.metadata.filePath,
          entry,
          metadata: this.metadata,
          ...this.logData,
        })
      } else {
        const foundFile = await FileModel.findOne({ path: this.updatedPath, complete: true })
        if (foundFile) {
          log.debug(
            { bucket: this.bucket, path: this.updatedPath, ...this.logData },
            'Skipping imported file as it has already been uploaded to S3.',
          )
          // auto-drain the stream
          stream.resume()
        } else {
          await putObjectStream(this.updatedPath, stream, this.bucket)
          await markFileAsCompleteAfterImport(this.updatedPath)
          log.debug(
            { bucket: this.bucket, path: this.updatedPath, name: entry.name, ...this.logData },
            'Imported file successfully uploaded to S3.',
          )
        }
        this.extractedFile = true
      }
    } else {
      // skip entry of type: link | symlink | directory | block-device | character-device | fifo | contiguous-file
      log.debug({ name: entry.name, type: entry.type, ...this.logData }, 'Skipping non-file entry.')
    }
  }

  // Type resolve
  finishListener(resolve: (reason?: FileMirrorInformation) => void, _reject: (reason?: unknown) => void) {
    super.finishListener(resolve, _reject)
  }
}
