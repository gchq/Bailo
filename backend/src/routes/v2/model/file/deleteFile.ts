import { Request, Response } from 'express'
import { z } from 'zod'

import { AuditInfo } from '../../../../connectors/audit/Base.js'
import audit from '../../../../connectors/audit/index.js'
import { removeFile } from '../../../../services/file.js'
import { registerPath } from '../../../../services/specification.js'
import { parse } from '../../../../utils/validate.js'

export const deleteFileSchema = z.object({
  params: z.object({
    modelId: z.string(),
    fileId: z.string(),
  }),
})

registerPath({
  method: 'delete',
  path: '/api/v2/model/{modelId}/file/{fileId}',
  tags: ['file'],
  description: 'Delete a file from a model.',
  schema: deleteFileSchema,
  responses: {
    200: {
      description: 'A message confirming the removal of the file.',
      content: {
        'application/json': {
          schema: z.object({
            message: z.string().openapi({ example: 'Successfully removed file' }),
          }),
        },
      },
    },
  },
})

interface DeleteFileResponse {
  message: string
}

export const deleteFile = [
  async (req: Request, res: Response<DeleteFileResponse>): Promise<void> => {
    req.audit = AuditInfo.DeleteFile
    const {
      params: { modelId, fileId },
    } = parse(req, deleteFileSchema)

    await removeFile(req.user, modelId, fileId)

    await audit.onDeleteFile(req, modelId, fileId)

    res.json({
      message: 'Successfully removed file.',
    })
  },
]
