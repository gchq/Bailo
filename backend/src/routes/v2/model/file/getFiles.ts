import bodyParser from 'body-parser'
import { Request, Response } from 'express'
import { z } from 'zod'

import { AuditInfo } from '../../../../connectors/audit/Base.js'
import audit from '../../../../connectors/audit/index.js'
import { FileWithScanResultsInterface } from '../../../../models/File.js'
import { getFilesByModel } from '../../../../services/file.js'
import { fileWithScanInterfaceSchema, registerPath } from '../../../../services/specification.js'
import { parse } from '../../../../utils/validate.js'

export const getFilesSchema = z.object({
  params: z.object({
    modelId: z.string({
      required_error: 'Must specify model id as param',
    }),
  }),
})

registerPath({
  method: 'get',
  path: '/api/v2/model/{modelId}/files',
  tags: ['file'],
  description: 'Get all of the files associated with a model.',
  schema: getFilesSchema,
  responses: {
    200: {
      description: 'An array of file instances.',
      content: {
        'application/json': {
          schema: z.object({
            files: z.array(fileWithScanInterfaceSchema),
          }),
        },
      },
    },
  },
})

interface GetFilesResponse {
  files: Array<FileWithScanResultsInterface>
}

export const getFiles = [
  bodyParser.json(),
  async (req: Request, res: Response<GetFilesResponse>) => {
    req.audit = AuditInfo.ViewFiles
    const {
      params: { modelId },
    } = parse(req, getFilesSchema)

    const files = await getFilesByModel(req.user, modelId)
    await audit.onViewFiles(req, modelId, files)

    return res.json({
      files,
    })
  },
]
