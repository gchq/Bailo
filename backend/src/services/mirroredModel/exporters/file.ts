import { Readable } from 'node:stream'

import { FileAction } from '../../../connectors/authorisation/actions.js'
import authorisation from '../../../connectors/authorisation/index.js'
import { ScanState } from '../../../connectors/fileScanning/Base.js'
import scanners from '../../../connectors/fileScanning/index.js'
import { FileWithScanResultsInterface } from '../../../models/File.js'
import { ModelDoc } from '../../../models/Model.js'
import { UserInterface } from '../../../models/User.js'
import config from '../../../utils/config.js'
import { BadReq, Forbidden } from '../../../utils/error.js'
import { downloadFile } from '../../file.js'
import { MirrorKind, MirrorLogData } from '../mirroredModel.js'
import { addEntryToTarGzUpload, initialiseTarGzUpload } from '../tarball.js'
import { BaseExporter } from './base.js'

export class FileExporter extends BaseExporter {
  protected readonly file: FileWithScanResultsInterface

  constructor(user: UserInterface, model: ModelDoc, file: FileWithScanResultsInterface, logData: MirrorLogData) {
    super(user, model, logData)
    this.file = file
  }

  getFile() {
    return this.file
  }

  protected _init() {
    if (this.file.size > config.modelMirror.export.maxSize) {
      throw BadReq('Requested export is too large.', {
        size: this.file.size,
        maxSize: config.modelMirror.export.maxSize,
      })
    }

    if (scanners.info()) {
      if (!this.file.avScan || this.file.avScan.length === 0) {
        throw BadReq('The file is missing AV scan(s).', { filename: this.file.name, fileId: this.file.id })
      } else if (this.file.avScan.some((scanResult) => scanResult.state !== ScanState.Complete)) {
        throw BadReq('The file has incomplete AV scan(s).', { filename: this.file.name, fileId: this.file.id })
      } else if (this.file.avScan.some((scanResult) => scanResult.isInfected)) {
        throw BadReq('The file has failed AV scan(s).', { filename: this.file.name, fileId: this.file.id })
      }
    }
  }

  protected async _checkAuths() {
    const fileAuth = await authorisation.file(this.user, this.model!, this.file, FileAction.Download)
    if (!fileAuth.success) {
      throw Forbidden(fileAuth.info, {
        userDn: this.user.dn,
        modelId: this.model.id,
        fileId: this.file.id,
        ...this.logData,
      })
    }
  }

  protected getInitialiseTarGzUploadParams(): Parameters<typeof initialiseTarGzUpload> {
    return [
      `${this.file.id}.tar.gz`,
      {
        schemaVersion: 1,
        exporter: this.user.dn,
        sourceModelId: this.model.id,
        mirroredModelId: this.model.settings.mirror.destinationModelId!,
        filePath: this.file.id,
        importKind: MirrorKind.File,
      },
      this.logData,
    ]
  }

  protected async _addData() {
    await addEntryToTarGzUpload(
      this.tarStream!,
      {
        type: 'stream',
        filename: this.file.id,
        stream: (await downloadFile(this.user, this.file.id)).Body as Readable,
        size: this.file.size,
      },
      this.logData,
    )
  }
}
