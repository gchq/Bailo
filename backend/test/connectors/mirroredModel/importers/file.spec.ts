import { PassThrough } from 'node:stream'

import { Headers } from 'tar-stream'
import { beforeEach, describe, expect, test, vi } from 'vitest'

import { FileImporter, FileMirrorMetadata } from '../../../../src/connectors/mirroredModel/importers/file.js'
import { MirrorKind } from '../../../../src/connectors/mirroredModel/index.js'

vi.mock('../../../../src/utils/config.js', () => ({
  __esModule: true,
  default: configMocks,
}))

const authMocks = vi.hoisted(() => ({
  default: {
    releases: vi.fn(),
  },
}))
vi.mock('../../../../src/connectors/authorisation/index.js', () => authMocks)

const configMocks = vi.hoisted(() => ({
  s3: {
    buckets: {
      uploads: 'uploads-bucket',
    },
  },
  modelMirror: {
    contentDirectory: 'content-dir',
    export: { concurrency: 1 },
  },
}))
vi.mock('../../../../src/utils/config.js', () => ({
  __esModule: true,
  default: configMocks,
}))

const logMocks = vi.hoisted(() => ({
  debug: vi.fn(),
}))
vi.mock('../../../../src/services/log.js', () => ({
  default: logMocks,
}))

const s3Mocks = vi.hoisted(() => ({
  putObjectStream: vi.fn(),
}))
vi.mock('../../../../src/clients/s3.js', () => s3Mocks)

const fileModelMocks = vi.hoisted(() => ({
  findOne: vi.fn(),
}))
vi.mock('../../../../src/models/File.js', () => ({
  __esModule: true,
  default: fileModelMocks,
}))

const fileServiceMocks = vi.hoisted(() => ({
  createFilePath: vi.fn(() => 'updated/file/path'),
  markFileAsCompleteAfterImport: vi.fn(),
}))
vi.mock('../../../../src/services/file.js', () => fileServiceMocks)

const registryMocks = vi.hoisted(() => ({
  joinDistributionPackageName: vi.fn(() => 'repo/path:tag'),
}))
vi.mock('../../../../src/services/registry.js', () => registryMocks)

const mockMetadata: FileMirrorMetadata = {
  importKind: MirrorKind.File,
  mirroredModelId: 'model123',
  filePath: 'original/file/path',
} as FileMirrorMetadata

describe('connectors > mirroredModel > importers > FileImporter', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  test('constructor > success', () => {
    const importer = new FileImporter(mockMetadata)

    expect(importer).toMatchSnapshot()
  })

  test('constructor > error importKind not File', () => {
    const badMetadata = { ...mockMetadata, importKind: 'OtherKind' } as any

    expect(() => new FileImporter(badMetadata)).toThrowError(
      /^Cannot parse compressed File: incorrect metadata specified./,
    )
  })

  test('processEntry > success upload new file to S3', async () => {
    fileModelMocks.findOne.mockResolvedValue(null)

    const importer = new FileImporter(mockMetadata)
    const entry: Headers = { name: 'file1', type: 'file' } as Headers
    const stream = new PassThrough()
    stream.end('file-contents')

    await importer.processEntry(entry, stream)

    expect(s3Mocks.putObjectStream).toHaveBeenCalledWith('updated/file/path', stream, configMocks.s3.buckets.uploads)
    expect(fileServiceMocks.markFileAsCompleteAfterImport).toHaveBeenCalledWith('updated/file/path')
    expect(importer).toMatchSnapshot()
  })

  test('processEntry > success skip already existing file', async () => {
    fileModelMocks.findOne.mockResolvedValue({ id: 'existingId' })

    const importer = new FileImporter(mockMetadata)
    const entry: Headers = { name: 'file1', type: 'file' } as Headers
    const stream = new PassThrough()
    stream.end('file-contents')
    // spy on resume
    const resumeSpy = vi.spyOn(stream, 'resume')

    await importer.processEntry(entry, stream)

    expect(resumeSpy).toHaveBeenCalled()
    expect(s3Mocks.putObjectStream).not.toHaveBeenCalled()
    expect(importer).toMatchSnapshot()
  })

  test('processEntry > error on multiple files', async () => {
    fileModelMocks.findOne.mockResolvedValue(null)

    const importer = new FileImporter(mockMetadata)
    const entry: Headers = { name: 'file1', type: 'file' } as Headers
    const stream = new PassThrough()
    stream.end('file-contents')
    const stream2 = new PassThrough()
    stream2.end('file-contents')

    await importer.processEntry(entry, stream)
    await expect(importer.processEntry(entry, stream2)).rejects.toThrowError(
      /^Cannot parse compressed file: multiple files found./,
    )
  })

  test('processEntry > success skip non-file entries', async () => {
    const importer = new FileImporter(mockMetadata)
    const entry: Headers = { name: 'dir', type: 'directory' } as Headers
    const stream = new PassThrough()

    await importer.processEntry(entry, stream)

    expect(logMocks.debug).toHaveBeenCalledWith({ name: 'dir', type: 'directory' }, 'Skipping non-file entry.')
  })

  test('finishListener > success calls BaseImporter behaviour', () => {
    const importer = new FileImporter(mockMetadata)
    const resolve = vi.fn()
    const reject = vi.fn()

    importer.finishListener(resolve, reject)

    expect(resolve).toHaveBeenCalledWith({ metadata: mockMetadata })
  })
})
