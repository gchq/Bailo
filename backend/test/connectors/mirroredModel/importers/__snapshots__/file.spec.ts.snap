// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`connectors > mirroredModel > importers > FileImporter > constructor > success 1`] = `
FileImporter {
  "bucket": "uploads-bucket",
  "extractedFile": false,
  "logData": undefined,
  "metadata": {
    "filePath": "original/file/path",
    "importKind": "file",
    "mirroredModelId": "model123",
  },
  "updatedPath": "updated/file/path",
}
`;

exports[`connectors > mirroredModel > importers > FileImporter > processEntry > success skip already existing file 1`] = `
FileImporter {
  "bucket": "uploads-bucket",
  "extractedFile": true,
  "logData": undefined,
  "metadata": {
    "filePath": "original/file/path",
    "importKind": "file",
    "mirroredModelId": "model123",
  },
  "updatedPath": "updated/file/path",
}
`;

exports[`connectors > mirroredModel > importers > FileImporter > processEntry > success upload new file to S3 1`] = `
FileImporter {
  "bucket": "uploads-bucket",
  "extractedFile": true,
  "logData": undefined,
  "metadata": {
    "filePath": "original/file/path",
    "importKind": "file",
    "mirroredModelId": "model123",
  },
  "updatedPath": "updated/file/path",
}
`;

exports[`services > mirroredModel > importers > FileImporter > constructor > success 1`] = `
FileImporter {
  "bucket": "uploads-bucket",
  "extractedFile": false,
  "logData": undefined,
  "metadata": {
    "filePath": "original/file/path",
    "importKind": "file",
    "mirroredModelId": "model123",
  },
  "updatedPath": "updated/file/path",
}
`;

exports[`services > mirroredModel > importers > FileImporter > processEntry > success skip already existing file 1`] = `
FileImporter {
  "bucket": "uploads-bucket",
  "extractedFile": true,
  "logData": undefined,
  "metadata": {
    "filePath": "original/file/path",
    "importKind": "file",
    "mirroredModelId": "model123",
  },
  "updatedPath": "updated/file/path",
}
`;

exports[`services > mirroredModel > importers > FileImporter > processEntry > success upload new file to S3 1`] = `
FileImporter {
  "bucket": "uploads-bucket",
  "extractedFile": true,
  "logData": undefined,
  "metadata": {
    "filePath": "original/file/path",
    "importKind": "file",
    "mirroredModelId": "model123",
  },
  "updatedPath": "updated/file/path",
}
`;
